### DISCUSSION TITLE
Easiest way to update the Stig Manager version to latest using docker compile?

### DISCUSSION THREAD

**Discussion by st3v1s** [NUWCDIVNPT/stig-manager#1208] [General]
Our users have requested that the lastest version of Stig Manager be available as they would like to take advantage of new features.  Our current build is version 1.3.7.  We have a docker-compile.yml wih image nuwcdivnpt/stig-manager:latest-ironbank.

Do I just need to do docker pull nuwcdivnpt/stig-manager:latest-ironbank to pull latest and start up containers?  Also I don't want to lose any data so I would do a docker stop, and then docker start.

Thoughts?

**Comment by cd-rite** [comment]
Hi @st3v1s Yes, it should be that simple!  The new version of the API container will make any required changes to the database.  It is always best to be sure you have backups of your Database in case there are any issues!

**Reply by st3v1s** [reply]
Thanks for the quick response.  One other question, what would be best method to make backup of the database (container)?

**Reply by cd-rite** [reply]
Well, I wouldn't say there is a  best way... it would depend on your organizational requirements for managing backups, etc..  But the **easiest** way would be to just stop the container and make a copy of the data directory from whatever volume you have it mounted to.

**Reply by st3v1s** [reply]
So yesterday I was able to pull the latest-ironbank stig manager but when I ran docker compose up, I lost all my data in keycloak and stig manager.  Next thing I was going to try was , docker compose stop, remove the stigman container, and then docker compose up, hoping that the user and other settings stay intact.  Thoughts?

Our users said that they can export the stigs in the collections so that they can import them into the new version of stig manager container....do you think that would be good work around?

**Reply by st3v1s** [reply]
So if I copy off the data directory, how can I apply it to the new image of stig manager that I have already done a docker pull on?  Is there an import option of sort?  Sorry I'm not a pro with docker yet.

**Reply by cd-rite** [reply]
The STIG Manager container doesn't store any data, it expects the database to be present as a resource for it to store its data.  If you back up the database data directory, those files just need to be provided to the database container for the data to persist.  This is done by mapping "volumes" with Docker, they provide a way to map data from the host system to the containers they are running.  I'd look into those with your setup, since it sounds like your docker-compose is not configured to use them...? Otherwise, when you brought up the new docker-compose orchestration, if the database was pointing at the same volume, (as well as keycloak stuff), the data should be there.

**Reply by st3v1s** [reply]
Ok that makes sense to me. And yes the current docker-compose was not configured to use volumes.  

So if I just tar up the _data director in /var/lib/docker/volumes/volumename/_data/    what command do I run to provide that tar file at the next docker compose build?

**Reply by cd-rite** [reply]
The tar would be your backup, it would need to be un-tarred and put wherever your configuration would look for it.  
Sorry, but this is really getting beyond what we can support regarding our app, and is more about general Docker usage, etc... I'd look here to see if you can get things the way you like them: https://docs.docker.com/storage/volumes/#use-a-volume-with-docker-compose

You can always un-tar the data into a new place, and bring up a new orchestration pointing at it to verify that it works. 

If you don't currently have your database using a volume, then your only option may be to export your .ckls as they stand in your existing deployment, and then re-import them into a new orchestration that is configured to use volumes, etc.  
Some pieces of our sample orchestration may be helpful:
https://github.com/NUWCDIVNPT/stigman-orchestration/blob/d2dae23c7edba3688dab7b1e02f7409e3b36d5f4/docker-compose.yml#L63

**Reply by st3v1s** [reply]
What do you mean by this statement: If you don't currently have your database using a volume, then your only option may be to export your .ckls as they stand in your existing deployment?   What are the .ckls files you are referring to?

I'm still testing different scenarios for this issue.

**Reply by cd-rite** [reply]
HI @st3v1s I was just describing an absolute last resort option if you can't get your database to persist data. You had said earlier:
> Our users said that they can export the stigs in the collections so that they can import them into the new version of stig manager container....do you think that would be good work around?

Those exports are usually in ".ckl" format.  They could be exported for each Collection in the app, and then reimported into a new instance with new Collections. It's not ideal, though: they would no longer have Review history data, user assignments, reference STIGs would have to be re-imported, etc.

**Reply by st3v1s** [reply]
Ohhhh, that's what you meant by .ckls!!  :-)

**Comment by st3v1s** [comment]
Finally solved my issue.  Was able to get Stigman Manager updated to 1.4.2 and use my historical data.  Also created persistent volume.

**Reply by cd-rite** [reply]
That's great! Glad you got it sorted out!
